### ISSUE TITLE
Scheduler memory keeps piling up after worker reconnecting (Windows platform)

### ISSUE THREAD

**Issue by chunmingchen** [dask/distributed#4571]
I have a distributed Dask cluster of around 50 machines where Dask workers constantly connects and disconnects to my scheduler, due to local activity of other processes on each machine. I found the scheduler quickly piles up memory usage when worker connects, but never releases memory after worker disconnection. The memory usage of the scheduler can stack up to 10GB, just after half a day. I had a run of 3 days, where the scheduler drained all system memory of 32GB.

To find the essential problem, I started a new conda env with just python==3.7 and latest dask distributed installed. Then I ran
- `dask-scheduler`
- `dask-worker localhost:8786`
- Kill the worker, and repeat launching the worker.
The memory of the scheduler could simply be seen going up.

Please advice on the issue and work around.  If some events or logs was intentionally kept in scheduler memory, is there any way to GC them?
 
**Environment**:
- Dask version: 2021.3.0
- Python version: 3.7.0
- Operating System: Both Windows 10 and Windows 7 were tested, giving same issue
- Install method (conda, pip, source): conda with channel conda-forge, clean install

- All packages:
 Name                    Version                   Build  Channel
bokeh                     2.3.0            py37h03978a9_0    conda-forge
certifi                   2020.12.5        py37h03978a9_1    conda-forge
click                     7.1.2              pyh9f0ad1d_0    conda-forge
cloudpickle               1.6.0                      py_0    conda-forge
cytoolz                   0.11.0           py37hcc03f2d_3    conda-forge
dask                      2021.3.0           pyhd8ed1ab_0    conda-forge
dask-core                 2021.3.0           pyhd8ed1ab_0    conda-forge
distributed               2021.3.0         py37h03978a9_0    conda-forge
freetype                  2.10.4               h546665d_1    conda-forge
fsspec                    0.8.7              pyhd8ed1ab_0    conda-forge
heapdict                  1.0.1                      py_0    conda-forge
intel-openmp              2020.3             h57928b3_311    conda-forge
jinja2                    2.11.3             pyh44b312d_0    conda-forge
jpeg                      9d                   h8ffe710_0    conda-forge
lcms2                     2.12                 h2a16943_0    conda-forge
libblas                   3.9.0                     8_mkl    conda-forge
libcblas                  3.9.0                     8_mkl    conda-forge
liblapack                 3.9.0                     8_mkl    conda-forge
libpng                    1.6.37               h1d00b33_2    conda-forge
libtiff                   4.2.0                hc10be44_0    conda-forge
locket                    0.2.0                      py_2    conda-forge
lz4-c                     1.9.3                h8ffe710_0    conda-forge
markupsafe                1.1.1            py37hcc03f2d_3    conda-forge
mkl                       2020.4             hb70f87d_311    conda-forge
msgpack-python            1.0.2            py37h8c56517_1    conda-forge
numpy                     1.20.1           py37hd20adf4_0    conda-forge
olefile                   0.46               pyh9f0ad1d_1    conda-forge
packaging                 20.9               pyh44b312d_0    conda-forge
pandas                    1.2.3            py37h08fd248_0    conda-forge
partd                     1.1.0                      py_0    conda-forge
pillow                    8.1.2            py37h96663a1_0    conda-forge
pip                       21.0.1           py37haa95532_0
psutil                    5.8.0            py37hcc03f2d_1    conda-forge
pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
python                    3.7.0                hea74fb7_0
python-dateutil           2.8.1                      py_0    conda-forge
python_abi                3.7                     1_cp37m    conda-forge
pytz                      2021.1             pyhd8ed1ab_0    conda-forge
pyyaml                    5.4.1            py37hcc03f2d_0    conda-forge
setuptools                52.0.0           py37haa95532_0
six                       1.15.0             pyh9f0ad1d_0    conda-forge
sortedcontainers          2.3.0              pyhd8ed1ab_0    conda-forge
tblib                     1.6.0                      py_0    conda-forge
tk                        8.6.10               h8ffe710_1    conda-forge
toolz                     0.11.1                     py_0    conda-forge
tornado                   6.1              py37hcc03f2d_1    conda-forge
typing_extensions         3.7.4.3                    py_0    conda-forge
vc                        14.2                 h21ff451_1
vs2015_runtime            14.27.29016          h5e58377_2
wheel                     0.36.2             pyhd3eb1b0_0
wincertstore              0.2                      py37_0
xz                        5.2.5                h62dcd97_1    conda-forge
yaml                      0.2.5                he774522_0    conda-forge
zict                      2.0.0                      py_0    conda-forge
zlib                      1.2.11            h62dcd97_1010    conda-forge
zstd                      1.4.9                h6255e5f_0    conda-forge

Update: 
This behavior also occurs on the one-year-older version distributed==2.12
After I killed and reconnected the worker for tens of times and waited for a few minutes for system settling down, the scheduler's memory usage never went back to the original.

**Comment by quasiben** [comment]
There have been reported memory leaks in the past but these, I recall,  being related to something deeper in within Python/NumPy.  I am surprised that starting the worker and closing it results in a memory leak on the scheduler.  Could I ask you to build a small reproducer to confirm this is happening ?

Something you could experiment with is calling the garbage collector on the scheduler:

```python
import gc
...
client.run_on_scheduler(gc.collect)
```

**Comment by chunmingchen** [comment]
quasiben:

Thanks for providing this quick test. On the clean env the gc command did lowered some amount of memory usage but did not go back to the original.

I have also tried GC'ing on my working scheduler which used 14GB of RAM at the moment. Only a fraction (less than 100MB) were released. Note, there was no worker/scheduler plugin used or any job running.

**Comment by chunmingchen** [comment]
I hope someone could help verify this issue on other OS platforms to identify whether this is just Windows problem.

**Comment by chrisroat** [comment]
I also see scheduler memory growth in my distributed application, and have to restart my clusters every few runs.  I find after a few GB of growth, the scheduler starts to get bogged down with CPU near 100% and things get flaky.

I have a GKE environment with preemptible nodes, and I use auto-scaling, so have a lot of disconnect/reconnect events.  In my past investigations, I could see memory increase even without disconnect/reconnect (#3898).

**Comment by chunmingchen** [comment]
I have conceived an interesting way to profile memory usage of a scheduler to find memory leak. Using standard library [tracemalloc](https://docs.python.org/3/library/tracemalloc.html), from client we can send two consecutive tasks of `client.run_on_scheduler` where `tracemalloc.take_snapshot()` is executed and collected. Then we can `compare` the two snapshots.

The client code looks like:
```python
def clear_logs(dask_scheduler): # As suggested in #3898 
    dask_scheduler.log.clear()
    dask_scheduler.transition_log.clear()
    dask_scheduler.events.clear()

def test_scheduler():
    import tracemalloc, time, gc
    client.run_on_scheduler(tracemalloc.start)
    print(client.run_on_scheduler(gc.collect)) # GC!
    snapshot1 = client.run_on_scheduler(tracemalloc.take_snapshot)
    time.sleep(100)
    print('stopping') # Notice not to test more worker connections/disconnections
    time.sleep(10)
    client.run_on_scheduler(clear_logs) # Clear logs!
    print(client.run_on_scheduler(gc.collect)) # GC!
    snapshot2 = client.run_on_scheduler(tracemalloc.take_snapshot)
    client.run_on_scheduler(tracemalloc.stop) # Remember to stop tracing for a production scheduler

    top_stats = snapshot2.compare_to(snapshot1, 'lineno')

    print("[ Top 10 differences ]")
    for stat in top_stats[:10]:
        print(stat)
```
Then as of in my prev experiments, just start a clean scheduler, run the above client code and keep running and killing a worker during the profiling period. The result is below:
```
[ Top 10 differences ]
...\lib\site-packages\bokeh\document\document.py:1180: size=208 KiB (+208 KiB), count=1278 (+1278), average=167 B
...\lib\site-packages\distributed\profile.py:52: size=139 KiB (+138 KiB), count=907 (+904), average=157 B
...\lib\site-packages\distributed\protocol\core.py:223: size=124 KiB (+124 KiB), count=1560 (+1554), average=82 B
...\lib\site-packages\distributed\scheduler.py:713: size=66.2 KiB (+66.2 KiB), count=808 (+808), average=84 B
...\lib\site-packages\distributed\profile.py:116: size=62.0 KiB (+61.5 KiB), count=599 (+595), average=106 B
...\lib\site-packages\distributed\profile.py:115: size=60.4 KiB (+59.9 KiB), count=584 (+580), average=106 B
...\lib\site-packages\distributed\profile.py:67: size=60.0 KiB (+59.8 KiB), count=786 (+784), average=78 B
...\lib\site-packages\distributed\profile.py:72: size=56.2 KiB (+56.0 KiB), count=546 (+544), average=105 B
...\lib\site-packages\tornado\platform\asyncio.py:215: size=34.4 KiB (+34.3 KiB), count=248 (+247), average=142 B
...\lib\site-packages\distributed\protocol\serialize.py:471: size=34.0 KiB (+34.0 KiB), count=296 (+296), average=118 B
```
**(Scheduler version: 2021.01.1)**

So maybe this is a good starting point to look into the problem.

**Comment by chrisroat** [comment]
Is it possible for you to check your scheduler growth on the system viewer with the current HEADs of dask and distributed?  In my tests (slightly different than yours), I've seen 3x larger growth.

**Comment by jrbourbeau** [comment]
Cross-referencing https://github.com/dask/distributed/pull/4604 which makes the size of the events log configurable

**Comment by chunmingchen** [comment]
Thanks jrbourbeau and chisroat for the update. I will try out with the latest version.

If the key change is to reduce the deque size, probably I can just reduce the value in the code of the installed conda site-packages? 

I also agree that the deque in #4604 seems the cause to the increased memory usage. However, if in my case the memory usage could went to 10GB+, with maxlen=100000 of the deque, each event log takes 1MB+ in average. Not sure this is realistic, since I believe during my runs there were only connection and disconnection events.

**Comment by chunmingchen** [comment]
Hello,

Good news: After reducing log maxlen from 100K to 1K, the scheduler ends up with 1.2GB memory usage after running for around 3 days.  This is much lower than 10GB+ as what I got originally.

The setting changes include
* `distributed.scheduler.transition-log-length: 1000` (originally 100K), and
* directly altering the line `self.events = defaultdict(lambda: deque(maxlen=1000))` in scheduler.py (originally 100K). As in #4604, this will become `distributed.scheduler.transition-log-length` config in the future.

Tested with current release version 2021.3.0.
Not sure which of the above two configs is the major cause of the issue, but at least now the problem seems eased.
